src/chainlink/ChainlinkOracle.sol

```diff
@@ -0,0 +1,80 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.20;
+
+// types
+import {IERC20} from "forge-std/interfaces/IERC20.sol";
+import {AggregatorV3Interface} from "./AggregatorV3Interface.sol";
+import {IOracle} from "../interfaces/IOracle.sol";
+import {BoringERC20} from "../libraries/BoringERC20.sol";
+
+contract ChainlinkOracle is IOracle {
+    using BoringERC20 for IERC20; // handles non-standard ERC20 tokens
+
+    AggregatorV3Interface public immutable FEED;
+    address public immutable NUMERATOR_ASSET;
+    address public immutable DENOMINATOR_ASSET;
+    uint256 public immutable MAX_STALENESS;
+
+    uint256 internal immutable NUMERATOR_SCALAR;
+    uint256 internal immutable DENOMINATOR_SCALAR;
+    uint256 internal immutable FEED_SCALAR;
+
+    constructor(address _feed, address _numeratorAsset, address _denominatorAsset, uint256 _maxStaleness) {
+        FEED = AggregatorV3Interface(_feed);
+        NUMERATOR_ASSET = _numeratorAsset;
+        DENOMINATOR_ASSET = _denominatorAsset;
+        MAX_STALENESS = _maxStaleness;
+
+        NUMERATOR_SCALAR = 10 ** IERC20(_numeratorAsset).safeDecimals();
+        DENOMINATOR_SCALAR = 10 ** IERC20(_denominatorAsset).safeDecimals();
+        FEED_SCALAR = 10 ** AggregatorV3Interface(_feed).decimals(); // decimals is not optional for the aggregator
+    }
+
+    /// @inheritdoc IOracle
+    function valueOf(address base, address quote, uint256 baseAmount) external view returns (uint256) {
+        bool isForward = _getQueryDirection(base, quote);
+        uint256 answer = _getAnswer(base, quote);
+
+        if (isForward) {
+            return (baseAmount * answer * DENOMINATOR_SCALAR) / (NUMERATOR_SCALAR * FEED_SCALAR);
+        } else {
+            return (baseAmount * NUMERATOR_SCALAR * FEED_SCALAR) / (answer * DENOMINATOR_SCALAR);
+        }
+    }
+
+    /// @inheritdoc IOracle
+    function priceOf(address base, address quote) external view returns (uint256) {
+        bool isForward = _getQueryDirection(base, quote);
+        uint256 answer = _getAnswer(base, quote);
+
```

Comment:
This returns prices as the value of one asset unit in terms of the other, while the spec states the prices should always be FP18.